### PR TITLE
Feature/batch_request

### PR DESCRIPTION
**Pull request by rolandvarga**
Adapt JSON batching when retrieving metrics

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -133,8 +176,17 @@ func (c *Collector) Collect(ch chan<- prometheus.Metric) {
 			return
 		}
 
-		for _, resource := range filteredResources {
-			c.collectResource(ch, resource, metricsStr, resourceGroup.Aggregations)
+		for _, f := range filteredResources {
+			var resource resourceMeta
+			resource.metrics = metricsStr
+
+			if len(resourceGroup.Aggregations) > 0 {
```
Comment (vincentcreusotfx) [inline main.go]: Could we factorize this ... it seems the same lines of code as for all kind of filter

File: main.go [hunk 2]
```diff
@@ -99,23 +104,61 @@ func (c *Collector) collectResource(ch chan<- prometheus.Metric, resource string
 	}
 }
 
+func (c *Collector) batchCollectResources(ch chan<- prometheus.Metric, resources []resourceMeta) {
+	const batchSize = 20
```
Comment (vincentcreusotfx) [inline main.go]: can you put that in the file's const declarations ?

File: main.go [hunk 3]
```diff
@@ -99,23 +104,61 @@ func (c *Collector) collectResource(ch chan<- prometheus.Metric, resource string
...
+	// organize resources based on batchSize
+	var resourceBatch [][]resourceMeta
+	for batchSize < len(resources) {
+		resources, resourceBatch = resources[batchSize:], append(resourceBatch, resources[0:batchSize])
+	}
+	resourceBatch = append(resourceBatch, resources) // don't forget to add remainder resources
+
+	// collect metrics per batch
+	for _, batch := range resourceBatch {
+		urls := []string{}
```
Comment (vincentcreusotfx) [inline main.go]: My linter says replace by var urls []string

File: main.go [hunk 4]
```diff
@@ -99,23 +104,61 @@ func (c *Collector) collectResource(ch chan<- prometheus.Metric, resource string
...
+
+	// organize resources based on batchSize
+	var resourceBatch [][]resourceMeta
+	for batchSize < len(resources) {
+		resources, resourceBatch = resources[batchSize:], append(resourceBatch, resources[0:batchSize])
+	}
+	resourceBatch = append(resourceBatch, resources) // don't forget to add remainder resources
+
+	// collect metrics per batch
+	for _, batch := range resourceBatch {
```
Comment (vincentcreusotfx) [inline main.go]: why not do the job in the resourceBatch array creation for clause directly, without creating a temp array

**Review by vincentcreusotfx — COMMENTED** [review]
I would remove diagnostic.go and the shell script as it's just here for analysis purpose. It doesn't change the fact that some batch request will improve the quality

**Review by vincentcreusotfx — APPROVED** [review]

**Review by zopanix — APPROVED** [review]
Too bad we're still not using the azure go sdk for this, we're simply re-writing it

**Pull request merged**